File: problems/0239-sliding-window-maximum/python/solution.py
from collections import deque


def maxSlidingWindow(nums, k):
    """Sliding Window Maximum — https://leetcode.com/problems/sliding-window-maximum/"""
    dq = deque()  # indices, values decreasing front -> back
    res = []
    for i, n in enumerate(nums):
        while dq and nums[dq[-1]] < n:
            dq.pop()
        dq.append(i)
        if dq[0] <= i - k:
            dq.popleft()
        if i >= k - 1:
            res.append(nums[dq[0]])
    return res

File: problems/0239-sliding-window-maximum/typescript/solution.ts
// Sliding Window Maximum — https://leetcode.com/problems/sliding-window-maximum/
export function maxSlidingWindow(nums: number[], k: number): number[] {
  const dq: number[] = []; // indices, values decreasing front -> back
  const res: number[] = [];
  for (let i = 0; i < nums.length; i++) {
    while (dq.length > 0 && nums[dq[dq.length - 1]] < nums[i]) dq.pop();
    dq.push(i);
    if (dq[0] <= i - k) dq.shift();
    if (i >= k - 1) res.push(nums[dq[0]]);
  }
  return res;
}
